### DISCUSSION TITLE
Use the BusinessCentral.LinterCop with the Run-ALPipeline

### DISCUSSION THREAD

**Discussion by kevin-smet** [StefanMaron/BusinessCentral.LinterCop#242] [Q&A]
Hello, I'm trying to use the LinterCop with the Run-ALPipeline, but I'm not able to get it working. 
First, I created a PowerShell (based on your DownloadFile.ps1) to download the DLL from GitHub. Then I defined it as a "customCodeCops" in the settings.json file.
The pipelines sees the DLL, but I get a warning that "An instance of analyzer BusinessCentral.LinterCop.Design.Rule0016CheckForMissingCaptions cannot be created from c:\sources\scripts\Analyzers\BusinessCentral.LinterCop.dll: Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis".

This makes sense as the DLL doesn't end up in the **Analyzers** directory (where the other cops are located). 

This is a part of the compiler command: ".... /analyzer:C:\build\vsix\extension\bin\Analyzers\Microsoft.Dynamics.Nav.UICop.dll /analyzer:c:\sources\scripts\BusinessCentral.LinterCop.dll ..."

The PowerShell download script, downloads the DLL to the local drive of my build server, but it needs to end up in the C:\build\vsix\extension\bin\Analyzers within the container. I don't see an option to specify that somewhere within the Run-ALPipeline command.

Do you have some tips or examples for me. 
Thanks a lot.

Kevin

**Comment by rvanbekkum** [comment]
I have written down how we have integrated the BusinessCentral.LinterCop in our build pipelines using `Compile-AppInBcContainer`, but it works the same for the `Run-ALPipeline` command.
You can find it in the [Custom Code Analyzer in Build Pipelines](https://robvanbekkum.nl/additional-build-steps/#custom-code-analyzer-in-build-pipelines) section of this post named ["Catch more errors in your Business Central apps adding steps to your CI/CD pipelines!"](https://robvanbekkum.nl/additional-build-steps/).

Like @StefanMaron described in the BcContainerHelper section of his [README.md](https://github.com/StefanMaron/BusinessCentral.LinterCop#bccontainerhelper) there is a `-CustomCodeCops` parameter in recent versions of the BcContainerHelper module for the `Compile-AppInBcContainer` and `Run-ALPipeline` commands to which you can pass the path to the `BusinessCentral.LinterCop.dll` file.
But please be aware of the following note:
> Be aware though, the BusinessCentral.LinterCop.dll needs to be placed in a folder shared with the container.

Does this help you to resolve your issue with integrating LinterCop in your build pipelines? :)

**Reply by kevin-smet** [reply]
Thanks for your suggestion, but it still doesn't change my issue that I have here:

- I have a PowerShell script that downloads the LinterCop. That's not a problem. I can see in my pipeline that it ends up in:
D:\BuildAgent\_work\149\s\ExtensionName\LinterCop\BusinessCentral.LinterCop.dll
- I'm able to pass that one as a **customCodeCops** parameter to the **Run-AlPipeline** command. The script finds the DLL, but copies it from that local folder (D:\BuildAgent\_work\....) in the Docker container. And it ends up in the folder: **c:\sources\LinterCop\BusinessCentral.LinterCop.dll** (this is within Docker container)
- But all the other CodeCops (UICop, CodeCop, AppSourceCop) they are located within the folder **C:\build\vsix\extension\bin\Analyzers** (within Docker container)
- The pipeline gives a warning that it finds the BusinessCentral.LinterCop.dll, but that it:  **Could not load file or assembly 'Microsoft.Dynamics.Nav.CodeAnalysis, Version=10.1.10.52128, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified**
- The CodeAnalysis.dll is located within the **C:\build\vsix\extension\bin\Analyzers** folder, but the LinterCop is in **c:\sources\LinterCop\BusinessCentral.LinterCop.dll** (both locations within Docker)

So, I don't know what else I can do? I'm using the correct parameters for the Run-AlPipeline. Maybe need to ask Freddy?

**Reply by rvanbekkum** [reply]
Could it be that you need to specify to use the latest/marketplace VSIX with the appropriate parameter so that it takes the AL compiler VSIX from the marketplace? I know LinterCop is build against the latest AL compiler DLLs.
That might explain the warning you are getting. We have seen that same warning before when we were trying to use LinterCop with an old version of alc (e.g., the version of alc shipped with the bcartifact you are using).

**Reply by kevin-smet** [reply]
I can try that.. but that means it won't use the ALC shipped with the BC Artifact.. So, if I use this on the Next Minor, or worse on the Next Major, then the pipeline also fails. Because the Next Major needs the VSIX that is shipped with that artifact. 
I've send an email to Freddy, but I haven't got a response yet.

**Reply by rvanbekkum** [reply]
Yes, but I think that is how it is at the moment. What we are currently doing is:
* Use the marketplace VSIX for all 'normal'/'vCurrent' build pipelines and use the latest stable release of BusinessCentral.LinterCop.
* Use the VSIX shipped with the BC artifacts for all vNext (NextMinor/NextMajor) build pipelines and don't use the BusinessCentral.LinterCop.

The latter you could resolve by using a build of BusinessCentral.LinterCop that is build against the vNext VSIX (AL compiler for the NextMajor build, the one shipped with the artifact), but unfortunately that is not available in the releases provided by Stefañ Maron here on GitHub.
Therefore, we decided not to use the BusinessCentral.LinterCop for vNext builds anymore, also considering there are a large number of false positives when you are _simply_ using the latest version of BusinessCentral.LinterCop.

I know @StefanMaron is also cracking his brain on how this can be handled in the best way, without too much hassle.
Perhaps he has some further thoughts/updates for you, if you would really like to get it working for vNext builds as well without too much hassle.

**Reply by kevin-smet** [reply]
Ok, thanks Rob. I'll try with this option then. We will see in the future how things are changing. For now, it would be a great thing to have this in the daily and CI builds.
